File: week5/윤여준/행렬 테두리 회전하기.cpp
// 맞왜틀

#include <bits/stdc++.h>
using namespace std;

int board[101][101];

int solve(int y1, int x1, int y2, int x2) {
    int m = 1e9;
    
    int temp = board[y1][x1];
    
    for(int i = y1; i < y2; i++) {
        board[i][x1] = board[i+1][x2];
        m = min(m, board[i][x1]);
    }
    
    for(int i = x1; i < x2; i++) {
        board[y2][i] = board[y2][i+1];
        m = min(m, board[y2][i]);
    }
    
    for(int i = y2; i > y1; i--) {
        board[i][x2] = board[i-1][x2];
        m = min(m, board[i][x2]);
    }
    
    for(int i = x2; i > x1; i--) {
        board[y1][i] = board[y1][i-1];
        m = min(m, board[y1][i]);
    }
    
    board[y1][x1+1] = temp;
    m = min(m, board[y1][x1+1]);
        
    return m;
}

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    
    for(int i = 0; i < rows; i++)
        for(int j = 0; j < columns; j++)
            board[i][j] = i * rows + j + 1;
    
    for(auto v : queries)
        answer.push_back(solve(v[0]-1, v[1]-1, v[2]-1, v[3]-1));
    
    
    return answer;
}

File: week5/이다혜/행렬테두리회전하기.kt
import kotlin.math.min

class Solution {
    fun solution(rows: Int, columns: Int, queries: Array<IntArray>): IntArray {
        var answer = intArrayOf()
        var board: Array<IntArray> = Array(rows){i -> IntArray(columns){j -> i*columns + j + 1}}

        queries.forEach {
            val startX = it[0] - 1
            val startY = it[1] - 1
            val endX = it[2] - 1
            val endY = it[3] - 1

            val topArr = IntArray(endY - startY){diff -> board[startX][startY+diff]}
            val rightArr = IntArray(endX - startX){diff -> board[startX+diff][endY]}
            val bottomArr = IntArray(endY - startY){diff -> board[endX][endY - diff]}
            val leftArr = IntArray(endX - startX){diff -> board[endX-diff][startY]}

            var min = rows * columns

            topArr.forEachIndexed { diff, element ->
                board[startX][startY+diff+1] = element
                min = min(min, element)
            }

            rightArr.forEachIndexed { diff, element ->
                board[startX+diff+1][endY] = element
                min = min(min, element)
            }

            bottomArr.forEachIndexed { diff, element ->
                board[endX][endY-diff-1] = element
                min = min(min, element)
            }

            leftArr.forEachIndexed { diff, element ->
                board[endX-diff-1][startY] = element
                min = min(min, element)
            }

            answer += min
        }

        return answer
    }
}

fun main(args: Array<String>) {
    val queries: Array<IntArray> = Array(3){ IntArray(4) }
    queries[0][0] = 2
    queries[0][1] = 2
    queries[0][2] = 5
    queries[0][3] = 4

    queries[1][0] = 3
    queries[1][1] = 3
    queries[1][2] = 6
    queries[1][3] = 6

    queries[2][0] = 5
    queries[2][1] = 1
    queries[2][2] = 6
    queries[2][3] = 3

    Solution().solution(6, 6, queries).forEach { println(it) }
}